### PR TITLE
Rule S4039: Interface methods should be callable by derived types

### LINKED ISSUE #434: Rule S4039: Interface methods should be callable by derived types
[RSPEC-4039](https://jira.sonarsource.com/browse/RSPEC-4039)

When a base type explicitly implements a public interface method, that method is only accessible, in derived type, through a reference to the current instance (namely this). If the derived type explicitly overrides that interface method, the base implementation becomes un-accessible.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #434

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                    ArePropertiesEquivalent),
+                SyntaxKind.PropertyDeclaration);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c => ReportOnIssue<EventDeclarationSyntax>(c, m => m.ExplicitInterfaceSpecifier, m => m.Identifier,
+                    AreEventsEquivalent),
+                SyntaxKind.EventDeclaration);
+        }
+
+        private static void ReportOnIssue<TMemberSyntax>(SyntaxNodeAnalysisContext analysisContext,
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs]: Yes that sucks but I tried 5 different approaches and this is the best. Feel free to suggest if you have any idea.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class InterfaceMethodsShouldBeCallableByChildTypes : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4039";
+        private const string MessageFormat = "Make '{0}' sealed, change to a non explicit declaration or provide a " +
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs]: Would that fit if you do that (new line before "=")? If not, it's fine.

private const string MessageFormat 
 = "Make '{0}' sealed, change to a non explicit declaration or provide a new method exposing the functionality of '{1}'.";

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                    if (classSymbol == null ||
+                        classSymbol.IsSealed ||
+                        classSymbol.IsStatic ||
+                        !classSymbol.IsPublicApi() ||
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs]: For chass, isn't that equivalent to checking declared accessibility?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                        .ToList()
+                        .ForEach(diagnostic => c.ReportDiagnostic(diagnostic));
+                }, SyntaxKind.ClassDeclaration);
+        }
+
+        private static bool IsTrackedMember(ISymbol symbol)
+        {
+            return symbol.Kind == SymbolKind.Event ||
+                symbol.Kind == SymbolKind.Property ||
+                (symbol.Kind == SymbolKind.Method && symbol.Name != ".ctor");
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs]: Perhaps use "IsConstructor"?
What happens for destructor?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        private static Diagnostic CreateDiagnostic(ClassDeclarationSyntax classDeclaration, ISymbol faultyMember)
+        {
+            var syntaxNode = faultyMember.DeclaringSyntaxReferences.First().GetSyntax();
+
+            Location location;
+            switch (syntaxNode.Kind())
+            {
+                case SyntaxKind.MethodDeclaration:
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs]: We could create a helper method here and reuse it in ParametersCorrectOrder.cs

Identifier GetIdentifier(SyntaxNode)

File: sonaranalyzer-dotnet/its/expected/akka.net/Akka.TestKit-{0D3CBAD0-BBDB-43E5-AFC4-ED1D3ECDC224}-S4039.json [hunk 6]
```diff
@@ -0,0 +1,329 @@
+{
+"issues":  [
+{
+"id":  "S4039",
+"message":  "Make 'TestActorRefBase' sealed, change to a non explicit declaration or provide a new method exposing the functionality of 'ISurrogated.ToSurrogate'.",
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/its/expected/akka.net/Akka.TestKit-{0D3CBAD0-BBDB-43E5-AFC4-ED1D3ECDC224}-S4039.json]: Quite a few reports on test files. Do we actually care if the method is not accessible in test? I'd say it doesn't matter and we should ignore them.
WDYT?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class InterfaceMethodsShouldBeCallableByChildTypes : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4039";
+        private const string MessageFormat = "Make '{0}' sealed, change to a non explicit declaration or provide a " +
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs]: I'd add a dash in "non-explicit", but perhaps check with Ann.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs (hunk 1 above)
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs]: I think I liked the previous version better. Let's have a chat.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs (hunk 2 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InterfaceMethodsShouldBeCallableByChildTypes.cs]: Too long.

File: sonaranalyzer-dotnet/its/expected/akka.net/Akka.TestKit-{0D3CBAD0-BBDB-43E5-AFC4-ED1D3ECDC224}-S4039.json (hunk 6 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/expected/akka.net/Akka.TestKit-{0D3CBAD0-BBDB-43E5-AFC4-ED1D3ECDC224}-S4039.json]: Good catch

**Pull request merged**